src/xml_parser_main_metadata_together.c

```diff
@@ -337,133 +328,217 @@ int cr_xml_parse_main_metadata_together(const char *primary_path,
                                         gboolean allow_out_of_order,
                                         GError **err)
 {
-    int ret = CRE_OK;
-    CR_FILE *primary_f = NULL;
-    CR_FILE *filelists_f = NULL;
-    CR_FILE *other_f = NULL;
-    GError *tmp_err = NULL;
+    assert(pkgcb || newpkgcb);
+    cr_PkgIterator* pkg_iterator = cr_PkgIterator_new(
+        primary_path, filelists_path, other_path, newpkgcb, newpkgcb_data, warningcb, warningcb_data, allow_out_of_order, err
+    );
 
-    cr_CbData cbdata;
-    cbdata.in_progress_pkgs_list = NULL;
-    cbdata.in_progress_pkgs_hash = NULL;
-    cbdata.newpkgcb = newpkgcb;
-    cbdata.newpkgcb_data = newpkgcb_data;
-    cbdata.pkgcb = pkgcb;
-    cbdata.pkgcb_data = pkgcb_data;
+    // TODO: do this more properly
+    assert(pkg_iterator);
+    assert(!*err);
 
-    if (allow_out_of_order) {
-        cbdata.in_progress_pkgs_hash = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
+    cr_Package* package = NULL;
+    GError* tmp_err = pkg_iterator->tmp_err;
+
+    while (package = cr_PkgIterator_parse_next(pkg_iterator, err)) {
+        // call user package callback
+        if (pkgcb && pkgcb(package, pkgcb_data, &tmp_err)) {
+            if (tmp_err) {
+                g_propagate_prefixed_error(err, tmp_err, "Parsing interrupted: ");
+            } else {
+                g_set_error(err, ERR_DOMAIN, CRE_CBINTERRUPTED, "Parsing interrupted");
+            }
+            // cr_package_free(package)
+            cr_PkgIterator_free(pkg_iterator, err);
+            return CR_CB_RET_ERR;
```

Comment:
Should we return the error code here? `return (*err)->code`.
This returns from `cr_xml_parse_main_metadata_together()` not from a callback.